### PR TITLE
[BUG] "Clear data" button malfunctioning

### PR DESCRIPTION
**Pull request by Aitorbp**
## Related Issues
App: https://github.com/owncloud/android/issues/4309

- [x] Add changelog files for the fixed issues in folder changelog/unreleased. More info [here](https://github.com/owncloud/android/tree/master/changelog#create-changelog-items)
- [x] Add feature to Release Notes in `ReleaseNotesViewModel.kt` creating a new `ReleaseNote()` with String resources (if required)

_____

## QA

Reports:

- [X] (1) Accounts removed in migration https://github.com/owncloud/android/pull/4350#issuecomment-2031686479

Final behaviour:

https://github.com/owncloud/android/pull/4350#issuecomment-2036919590

### CODE REVIEW

**Review by JuancaG05 — CHANGES_REQUESTED** [review]
Some changes requested here @Aitorbp

File: changelog/unreleased/4350 [hunk 1]
```diff
@@ -0,0 +1,7 @@
+Bugfix: Clear data button enabled in the App info
+
+The clear data button have been enabled to delete the application data from the App info in the Settings app.
+Shared preferences, temporary files, accounts and the local database will be cleared when the button is pressed.
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply changelog/unreleased/4350]: Done

File: owncloudApp/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -211,10 +210,6 @@
                 android:name="android.app.searchable"
                 android:resource="@xml/users_and_groups_searchable" />
         </activity>
-        <activity
-            android:name=".ui.activity.ManageSpaceActivity"
-            android:label="@string/manage_space_title"
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudApp/src/main/AndroidManifest.xml:216]: Done

File: owncloudApp/src/main/res/layout/activity_manage_space.xml [hunk 3]
```diff
@@ -1,46 +0,0 @@
...
-
-    <TextView
-        android:id="@+id/general_description"
-        style="?android:attr/editTextPreferenceStyle"
-        android:layout_width="match_parent"
-        android:layout_height="wrap_content"
-        android:layout_gravity="center_vertical"
-        android:layout_margin="@dimen/standard_margin"
-        android:singleLine="false"
-        android:text="@string/manage_space_description" />
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudApp/src/main/res/layout/activity_manage_space.xml:32]: Deno

File: owncloudApp/src/main/java/com/owncloud/android/presentation/authentication/AccountUtils.java [hunk 4]
```diff
@@ -81,6 +83,14 @@ public static Account[] getAccounts(Context context) {
         return accountManager.getAccountsByType(MainApp.Companion.getAccountType());
     }
 
+    public static void deleteAccount(Context context) {
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/authentication/AccountUtils.java]: Yes, done

File: owncloudApp/src/main/res/values/strings.xml [hunk 5]
```diff
@@ -739,6 +739,8 @@
     <string name="release_notes_4_3_0_subtitle_5">The lens icon in the toolbar was removed when listing spaces in folder picker</string>
     <string name="release_notes_4_3_0_title_6">Add warning in http connections</string>
     <string name="release_notes_4_3_0_subtitle_6">Added a warning dialog in the login view when it is a http connection</string>
+    <string name="release_notes_4_3_0_title_7">Clear data button enabled in the App info</string>
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudApp/src/main/res/values/strings.xml]: Done

File: owncloudApp/src/main/res/values/strings.xml [hunk 6]
```diff
@@ -739,6 +739,8 @@
...
+    <string name="release_notes_4_3_0_subtitle_7">The clear data button have been enabled to delete the application data from the App info in the Settings app. The application will perform a hard reset if this button is pressed</string>
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudApp/src/main/res/values/strings.xml]: Done

File: owncloudApp/src/main/java/com/owncloud/android/MainApp.kt [hunk 7]
```diff
@@ -133,24 +135,30 @@ class MainApp : Application() {
...
-                            checkbox.setText(R.string.ocis_accounts_warning_checkbox_message)
-                            val builder = AlertDialog.Builder(activity).apply {
-                                setView(checkboxDialog)
-                                setTitle(R.string.ocis_accounts_warning_title)
-                                setMessage(R.string.ocis_accounts_warning_message)
-                                setCancelable(false)
-                                setPositiveButton(R.string.ocis_accounts_warning_button) { _, _ ->
-                                    if (checkbox.isChecked) {
-                                        pref.edit().putBoolean(PREFERENCE_KEY_DONT_SHOW_OCIS_ACCOUNT_WARNING_DIALOG, true).apply()
+                        val isClearDataClicked = pref.contains(PREFERENCE_CLEAR_DATA)
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudApp/src/main/java/com/owncloud/android/MainApp.kt]: Changed by: `isClearDataPrefAvailable`

File: owncloudApp/src/main/java/com/owncloud/android/MainApp.kt [hunk 8]
```diff
@@ -133,24 +135,30 @@ class MainApp : Application() {
...
+                        if (!isClearDataClicked) {
+                            AccountUtils.deleteAccount(appContext)
+                            WhatsNewActivity.runIfNeeded(activity)
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudApp/src/main/java/com/owncloud/android/MainApp.kt]: Changed the order

File: owncloudApp/src/main/res/values/strings.xml [hunk 9]
```diff
@@ -617,11 +617,6 @@
     <string name="action_upload_retry">Retry</string>
     <string name="action_upload_clear">Clear</string>
 
-    <string name="manage_space_title">Manage space</string>
-    <string name="manage_space_description">This feature is not currently available. Next versions will include new features for automatic and manual removal of storage. Stay tuned!</string>
-    <string name="manage_space_clear_data">Clear data</string>
-    <string name="manage_space_error">Some files could not be deleted.</string>
```
Comment (JuancaG05) [inline owncloudApp/src/main/res/values/strings.xml:623]: This is not enough, remember that when we remove a String, we need to remove it in every language...

File: owncloudApp/src/main/java/com/owncloud/android/MainApp.kt [hunk 10]
```diff
@@ -133,24 +135,30 @@ class MainApp : Application() {
...
-                            checkbox.setText(R.string.ocis_accounts_warning_checkbox_message)
-                            val builder = AlertDialog.Builder(activity).apply {
-                                setView(checkboxDialog)
-                                setTitle(R.string.ocis_accounts_warning_title)
-                                setMessage(R.string.ocis_accounts_warning_message)
-                                setCancelable(false)
-                                setPositiveButton(R.string.ocis_accounts_warning_button) { _, _ ->
-                                    if (checkbox.isChecked) {
-                                        pref.edit().putBoolean(PREFERENCE_KEY_DONT_SHOW_OCIS_ACCOUNT_WARNING_DIALOG, true).apply()
+                        val isClearDataPrefAvailable = pref.contains(PREFERENCE_CLEAR_DATA)
```
Comment (JuancaG05) [inline owncloudApp/src/main/java/com/owncloud/android/MainApp.kt:138]: This name doesn't indicate what does the fact that the `PREFERENCE_CLEAR_DATA` exists mean, will be harder for the future to understand it

File: owncloudApp/src/main/java/com/owncloud/android/MainApp.kt [hunk 11]
```diff
@@ -133,24 +135,30 @@ class MainApp : Application() {
...
                                     }
                                 }
+                                val alertDialog = builder.create()
+                                alertDialog.show()
                             }
-                            val alertDialog = builder.create()
-                            alertDialog.show()
+                        } else { // "Clear data" button is pressed from the app settings in the device settings.
+                            AccountUtils.deleteAccounts(appContext)
+                            WhatsNewActivity.runIfNeeded(activity)
```
Comment (JuancaG05) [inline owncloudApp/src/main/java/com/owncloud/android/MainApp.kt:161]: There was no need to change the order of this `if else` block

**Review by JuancaG05 — CHANGES_REQUESTED** [review]
Some more changes here @Aitorbp

File: owncloudApp/src/main/java/com/owncloud/android/MainApp.kt (hunk 10 above)
Comment (JuancaG05) [reply owncloudApp/src/main/java/com/owncloud/android/MainApp.kt:138]: `PREFERENCE_CLEAR_DATA` is not suitable at all either, what would you think that a preference called `PREFERENCE_CLEAR_DATA` set to `true` means?

File: owncloudApp/src/main/res/values/strings.xml (hunk 9 above)
Comment (Aitorbp) [reply owncloudApp/src/main/res/values/strings.xml:623]: Sorry...done

File: owncloudApp/src/main/java/com/owncloud/android/MainApp.kt (hunk 10 above)
Comment (Aitorbp) [reply owncloudApp/src/main/java/com/owncloud/android/MainApp.kt]: Changed

**Review by JuancaG05 — APPROVED** [review]
LGTM

**Comment by jesmrec** [comment]
About the removed translations, i think that removing the original string from `setup.xml` , transifex will remove it from all translation files... i'm not totally sure about this, any further idea?

**Comment by jesmrec** [comment]
### (1)

The solution is based in a preference `PREFERENCE_CLEAR_DATA_ALREADY_TRIGGERED` which existence deals some actions, f. ex. removing the existent accounts. The problem comes with the migrations. The preference did not exist in advance, therefore when migration takes part, the accounts are going to be removed. That's not expected and must be fixed.

**Comment by jesmrec** [comment]
OTOH, it'd be useful to detail in the issue the last status update. Last comment is about suggesting the account removal, but not the final decision and why.

**Comment by Aitorbp** [comment]
After the[ above investigation](https://github.com/owncloud/android/issues/4309#issuecomment-1994071157), it has been decided to clear the application's accounts when the user clicks on Clear data button. 

In order for the application to internally know that the Clear data button has been pressed, a `Shared Preferences` called `PREFERENCE_CLEAR_DATA_ALREADY_TRIGGERED` has been created. This variable will be given a value in the `FileDisplayActivity`, if it has no value it indicates that the preferences have been cleared, therefore a Clear data has been made. 

In the same way, the change of version of the application has been controlled in the Clear data condition, as it could be the case that if a user upgrades to a new version, in this new version, the Clear data preference will still not exist in this flow and therefore the accounts will be deleted. This is why the version change has been managed.

**Comment by jesmrec** [comment]
Checks performed:

- [X] Clear data with accounts in device: data cleaned up and accounts removed when app is reopen (not before)
- [X] Clear data with no accounts in device

Migrations:

**Release mode**: required change in version code, otherwise everything removed. We can assume that for release builds. 
**Debug mode**: works only for versions starting from this (containing the version code inside the shared preferences)

Go on with this. Will need another check in the release process.

**Pull request merged**